docs/guide/api-vite-environment.md

```diff
@@ -804,7 +804,74 @@ await createServer({
 
 :::
 
-`RemoteRunnerTransport` and `RemoteEnvironmentTransport` are meant to be used together. If you don't use either of them, then you can define your own function to communicate between the runner and the server.
+`RemoteRunnerTransport` and `RemoteEnvironmentTransport` are meant to be used together, but you don't have to use them at all. You can define your own function to communicate between the runner and the server. For example, if you connect to the environment via an HTTP request, you can call `fetch().json()` in `fetchModule` function:
+
+```ts
+import { ESModulesEvaluator, ModuleRunner } from 'vite/module-runner'
+
+const runner = new ModuleRunner(
+  {
+    root: fileURLToPath(new URL('./', import.meta.url)),
+    transport: {
+      async fetchModule(id, importer) {
+        const response = await fetch(
+          `http://my-vite-server/fetch?id=${id}&importer=${importer}`,
+        )
+        return response.json()
+      },
+    },
+  },
+  new ESModulesEvaluator(),
+)
+
+await runner.import('/entry.js')
+```
+
+::: warning Acessing Module on the Server
+We do not want to encourage communication between the server and the runner. One of the problems that were exposed with `vite.ssrLoadModule` is over-reliance on the server state inside the processed modules. This makes it harder to implement runtime-agnostic SSR since there might be no direct access to the server APIs that are available in Node.js.
+
+Instead, we recommend using virtual modules to import the state and process it inside the loaded environment module:
+
+```ts
+import { runner } from './ssr-module-runner.js'
+import { processRoutes } from './routes.js'
+
+const { routes } = await runner.import('ssr:routes')
+processRoutes(routes)
+```
+
+You can also use virtual modules to load HTML:
+
+```ts {13-21}
+function vitePluginVirtualIndexHtml(): Plugin {
+  let server: ViteDevServer | undefined
+  return {
+    name: vitePluginVirtualIndexHtml.name,
+    configureServer(server_) {
+      server = server_
+    },
+    resolveId(source, _importer, _options) {
+      return source === 'virtual:index-html' ? '\0' + source : undefined
+    },
+    async load(id, _options) {
+      if (id === '\0' + 'virtual:index-html') {
```

Comment:
Should we expose this as something like `runner.html(url, htmlPath)`? Feels weird that every framework now has to implement this

cc @patak-dev @hi-ogawa

Reply:
To support `runner.html()`, we would have to ask for `fetchHtml` method alongside `fetchModule` (right now transport only requires `fetchModule`)

Reply:
I use `index.html` for my quick setup (and also create-vite-extra template too e.g. https://github.com/bluwy/create-vite-extra/blob/master/template-ssr-react/server.js), but in practice, I'm not sure if actual framework would benefit from this since they usually have more opinionated way to generate the html shell.

At least for the frameworks I'm aware of, they don't have `index.html` convention. Instead, they have only "client entry" script and they can use it as `rollupOpitons.input`. They also have an access to client build manifest, so they can find "production client entry" asset path from manifest and renders `<script>` (also `<link>` for css, preload, etc...) on their own into their html shell.

So, I think having `runner.html()` would help only "entry level" ssr framework exploration and I'm not sure if it's worth it to have in core. For that case, showing this `virtual:index-html` idea here is probably helpful enough.